cmd/kubeadm/app/cmd/alpha/certs.go

```diff
@@ -82,9 +102,93 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
 	cmd.AddCommand(newCmdCertsRenewal(out))
 	cmd.AddCommand(newCmdCertsExpiration(out, constants.KubernetesDir))
 	cmd.AddCommand(NewCmdCertificateKey())
+	cmd.AddCommand(newCmdGenCSR())
+	return cmd
+}
+
+// genCSRConfig is the configuration required by the gencsr command
+type genCSRConfig struct {
+	kubeadmConfigPath string
+	certDir           string
+	kubeConfigDir     string
+	kubeadmConfig     *kubeadmapi.InitConfiguration
+}
+
+func newGenCSRConfig() *genCSRConfig {
+	return &genCSRConfig{
+		kubeConfigDir: kubeadmconstants.KubernetesDir,
+	}
+}
+
+func (o *genCSRConfig) addFlagSet(flagSet *pflag.FlagSet) {
+	options.AddConfigFlag(flagSet, &o.kubeadmConfigPath)
+	options.AddCertificateDirFlag(flagSet, &o.certDir)
+	options.AddKubeConfigDirFlag(flagSet, &o.kubeConfigDir)
+}
+
+// load merges command line flag values into kubeadm's config.
+// Reads Kubeadm config from a file (if present)
+// else use dynamically generated default config.
+// This configuration contains the DNS names and IP addresses which
+// are encoded in the control-plane CSRs.
+func (o *genCSRConfig) load() (err error) {
+	o.kubeadmConfig, err = configutil.LoadOrDefaultInitConfiguration(
+		o.kubeadmConfigPath,
+		&kubeadmapiv1beta2.InitConfiguration{},
+		&kubeadmapiv1beta2.ClusterConfiguration{},
+	)
+	if err != nil {
+		return err
+	}
+	// --cert-dir takes priority over kubeadm config if set.
+	if o.certDir != "" {
+		o.kubeadmConfig.CertificatesDir = o.certDir
+	}
+	return nil
+}
+
+// newCmdGenCSR returns cobra.Command for generating keys and CSRs
+func newCmdGenCSR() *cobra.Command {
+	config := newGenCSRConfig()
+
+	cmd := &cobra.Command{
+		Use:     "generate-csr",
+		Short:   "Generate keys and certificate signing requests",
+		Long:    generateCSRLongDesc,
+		Example: generateCSRExample,
+		Args:    cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := config.load()
+			if err != nil {
+				return err
+			}
+			return runGenCSR(config)
+		},
+	}
+	config.addFlagSet(cmd.Flags())
 	return cmd
 }
 
+// runGenCSR contains the logic of the generate-csr sub-command.
+func runGenCSR(config *genCSRConfig) error {
+	leafCertificates, err := certsphase.LeafCertificates(certsphase.GetDefaultCertList())
+	if err != nil {
+		return err
+	}
+	if err := certsphase.CreateKeyAndCSRFiles(config.kubeadmConfig, leafCertificates); err != nil {
+		return err
+	}
+
+	kubeConfigs, err := kubeconfigphase.GetDefaultKubeConfigs(config.kubeadmConfig)
+	if err != nil {
+		return err
+	}
+	if err := kubeconfigphase.CreateKubeConfigAndCSRFiles(config.kubeConfigDir, config.kubeadmConfig, kubeConfigs); err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
* moved all the certificate CSR code into `cmd/kubeadm/app/phases/certs` with unit tests
 * moved all the kubeconfig CSR code into `cmd/kubeadm/app/phases/kubeconfig` with unit tests
 * Tried to bring more consistency between these two packages, with both now having a  public `GetDefaultXXX` function. 
 * Both have a new `CreateXXXAndCSRFiles` function which accepts a list of things.
 * This function has a (slow) unit test for the happy path only, but each of the packages has more complete unit tests for error cases.